File: app/src/main/java/com/gitsearch/ui/BaseActivity.java
package com.gitsearch.ui;

import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import butterknife.ButterKnife;
import dagger.android.AndroidInjection;

public abstract class BaseActivity extends AppCompatActivity {

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    if (useInject()) {
      AndroidInjection.inject(this);
    }

    super.onCreate(savedInstanceState);
    setContentView(getLayoutRes());
    ButterKnife.bind(this);
  }

  protected boolean useInject() {
    return false;
  }

  @LayoutRes
  protected abstract int getLayoutRes();

}

File: app/src/main/java/com/gitsearch/ui/BaseActivity.kt
package com.gitsearch.ui

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import dagger.android.AndroidInjection
import dagger.android.DispatchingAndroidInjector
import dagger.android.support.HasSupportFragmentInjector
import javax.inject.Inject

abstract class BaseActivity(
        @LayoutRes private val layoutResId: Int
) : AppCompatActivity(), HasSupportFragmentInjector {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    @Inject
    lateinit var supportFragmentInjector: DispatchingAndroidInjector<Fragment>

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(layoutResId)
    }

    override fun supportFragmentInjector() = supportFragmentInjector

    protected inline fun <reified T : ViewModel> viewModel() = lazy<T> {  ViewModelProviders.of(this, viewModelFactory).get(T::class.java) }

}
